File: 알고리즘 문제풀이/kakao/다트게임.py
#S:1제곱
#D:2제곱
#T:3제곱
# *스타상:해당 점수와 바로 전에 얻은 점수 각 2배),
# #야차상:해당 점수 마이너스)
def solution(dartResult):
    answer,num = [],""
    for d in dartResult:
        if d=="S":
            if(num!=""):
                answer.append(int(num))
                num=""
        elif d=="D":
            if num!="":
                answer.append(int(num)**2)
                num=""
        elif d=="T":
            if num!="":
                answer.append(int(num)**3)
                num=""
        elif d=="*":
            if len(answer)>0:
                if len(answer)==1:
                    answer[-1]=answer[-1]*2
                elif len(answer)>1:
                    answer[-1]=answer[-1]*2
                    answer[-2]=answer[-2]*2
        elif d=="#":
            if len(answer)>0:
                answer[-1]=answer[-1]*-1
        else:
            num+=d
    return sum(answer)

File: 알고리즘 문제풀이/kakao/다트게임.js
function solution(dartResult) {
  let answer = 0,
    arr = [],
    number = "";
  for (let i = 0; i < dartResult.length; i++) {
    if (dartResult[i] === "S") {
      number = parseInt(number, 10);
      arr.push(number);
      number = "";
    } else if (dartResult[i] === "D") {
      number = parseInt(number, 10) * parseInt(number, 10);
      arr.push(number);
      number = "";
    } else if (dartResult[i] === "T") {
      number =
        parseInt(number, 10) * parseInt(number, 10) * parseInt(number, 10);
      arr.push(number);
      number = "";
    } else {
      if (dartResult[i] !== "*" && dartResult[i] !== "#") {
        number += dartResult[i];
      } else if (dartResult[i] === "*") {
        if (arr.length === 1) arr[0] *= 2;
        else if (arr.length >= 2) {
          arr[arr.length - 1] *= 2;
          arr[arr.length - 2] *= 2;
        }
      } else if (dartResult[i] === "#") {
        if (arr.length >= 1) arr[arr.length - 1] *= -1;
      }
    }
  }
  answer = arr.reduce((i, j) => i + j, 0);
  return answer;
}
console.log(solution("1S2D*3T"));
console.log(solution("1D2S#10S"));
console.log(solution("1D2S0T"));
console.log(solution("1S*2T*3S"));
